internal/util/metrics/metrics.go

```diff
@@ -4,114 +4,271 @@ import (
+		ConstLabels: defaultLabels,
+	}, []string{LabelK8SVersion})
+
+	// clusterCacheError count the total number of cluster cache errors
+	clusterCacheError = promauto.NewCounterVec(prometheus.CounterOpts{
+		Name:        "numaflow_cluster_cache_error_total",
+		Help:        "The total number of cluster cache error",
+		ConstLabels: defaultLabels,
+	}, []string{})
 )
 
 // RegisterCustomMetrics registers the custom metrics to the existing global prometheus registry for pipelines, ISB service and numaflow controller
 func RegisterCustomMetrics() *CustomMetrics {
-	metrics.Registry.MustRegister(pipelinesGauge, isbsvcGauge, numaflowControllerGauge)
+	metrics.Registry.MustRegister(pipelinesGauge, pipelineSynced, isbServiceGauge, numaflowControllerGauge, reconciliationDuration,
+		pipelineSyncFailed, pipelineRolloutQueueLength, isbServiceSyncFailed, numaflowControllerSyncFailed, numaflowKubeRequestCounter,
+		numaflowKubectlExecutionCounter, kubeResourceMonitored, kubeResourceCache, clusterCacheError)
 
 	return &CustomMetrics{
-		pipelinesGauge:          pipelinesGauge,
-		pipelineCounterMap:      make(map[string]struct{}),
-		isbsvcGauge:             isbsvcGauge,
-		isbsvcCounterMap:        make(map[string]struct{}),
-		numaflowControllerGauge: numaflowControllerGauge,
+		PipelinesGauge:                   pipelinesGauge,
+		PipelineCounterMap:               make(map[string]struct{}),
+		PipelineSynced:                   pipelineSynced,
+		PipelineSyncFailed:               pipelineSyncFailed,
+		PipelineRolloutQueueLength:       pipelineRolloutQueueLength,
+		ISBServiceGauge:                  isbServiceGauge,
+		ISBServiceCounterMap:             make(map[string]struct{}),
+		ISBServicesSynced:                isbServicesSynced,
+		ISBServiceSyncFailed:             isbServiceSyncFailed,
+		NumaflowControllerGauge:          numaflowControllerGauge,
+		NumaflowControllerVersionCounter: make(map[string]map[string]struct{}),
+		NumaflowControllerSynced:         numaflowControllerSynced,
+		NumaflowControllerSyncFailed:     numaflowControllerSyncFailed,
+		NumaflowKubeRequestCounter:       numaflowKubeRequestCounter,
+		NumaflowKubectlExecutionCounter:  numaflowKubectlExecutionCounter,
+		ReconciliationDuration:           reconciliationDuration,
+		KubeResourceMonitored:            kubeResourceMonitored,
+		KubeResourceCache:                kubeResourceCache,
+		ClusterCacheError:                clusterCacheError,
 	}
 }
 
 // IncPipelineMetrics increments the pipeline counter if it doesn't already know about the pipeline
 func (m *CustomMetrics) IncPipelineMetrics(name, namespace string) {
 	pipelineLock.Lock()
 	defer pipelineLock.Unlock()
-	m.pipelineCounterMap[name+"_"+namespace] = struct{}{}
-	m.pipelinesGauge.WithLabelValues().Set(float64(len(m.pipelineCounterMap)))
+	if _, ok := m.PipelineCounterMap[name+"_"+namespace]; !ok {
+		m.PipelineCounterMap[name+"_"+namespace] = struct{}{}
+		m.PipelineSynced.WithLabelValues().Inc()
```

Comment:
Compare this with `pipelinesyncfailed` which is called whenever there is a failure in sync reconciliation. this should not only increments if it is not a known pipeline.  Instead this should be increment in each sync reconciliation (fail or succeed)

Reply:
@xdevxy  so, you're saying you want "sync success" and "syncs failures" to both represent actual reconciliations. In that case we should change the description of pipelineSynced from "The total number of pipeline synced" to "the total number of successful syncs"?

Reply:
yeah it is the actual reconciliation. If we change to "the total number of successful syncs", then it includes only successful one, the current one covers all (successful or not). I'm fine with either ways, just the logic needs to be correct.

Reply:
I have updated it, and it’s covering all (successful or fail)